File: template-parts/code-library/snippets/go/circuit_breaker.go
// Package circuitbreaker implements the circuit breaker pattern for Go.
// Purpose: Prevent cascading failures by stopping requests to a failing service.
// Usage:
//   cb := circuitbreaker.New(breaker.Config{
//       MaxRequests: 3,
//       Interval:   30 * time.Second,
//       Timeout:    10 * time.Second,
//   })
//   err := cb.Execute(func() error { return callService() })
//
// Dependencies: standard library only
package circuitbreaker

import (
	"context"
	"errors"
	"sync"
	"time"
)

// State represents the circuit breaker state.
type State int

const (
	StateClosed State = iota
	StateOpen
	StateHalfOpen
)

func (s State) String() string {
	switch s {
	case StateClosed:
		return "closed"
	case StateOpen:
		return "open"
	case StateHalfOpen:
		return "half-open"
	default:
		return "unknown"
	}
}

// Errors returned by the circuit breaker.
var (
	ErrOpenState = errors.New("circuit breaker is open")
)

// Config configures the circuit breaker.
type Config struct {
	MaxRequests int           // max requests in half-open state before transitioning
	Interval    time.Duration // period of closed state to reset
	Timeout     time.Duration // duration of open state before half-open
}

// DefaultConfig returns sensible defaults.
func DefaultConfig() Config {
	return Config{
		MaxRequests: 1,
		Interval:   30 * time.Second,
		Timeout:    10 * time.Second,
	}
}

// CircuitBreaker implements the circuit breaker pattern.
type CircuitBreaker struct {
	config Config
	mu     sync.RWMutex
	state  State

	failures    int
	lastFailure time.Time

	openTime time.Time // when the circuit transitioned to open
}

// New creates a new circuit breaker.
func New(config Config) *CircuitBreaker {
	if config.MaxRequests <= 0 {
		config.MaxRequests = DefaultConfig().MaxRequests
	}
	return &CircuitBreaker{
		config: config,
		state:  StateClosed,
	}
}

// State returns the current circuit breaker state.
func (cb *CircuitBreaker) State() State {
	cb.mu.RLock()
	defer cb.mu.RUnlock()
	return cb.state
}

// Execute runs fn through the circuit breaker.
// Returns ErrOpenState if the circuit is open.
func (cb *CircuitBreaker) Execute(ctx context.Context, fn func() error) error {
	if !cb.allowRequest() {
		return ErrOpenState
	}

	err := fn()

	cb.recordResult(err)
	return err
}

// allowRequest checks if a request should be allowed through.
func (cb *CircuitBreaker) allowRequest() bool {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case StateClosed:
		return true

	case StateOpen:
		// Check if timeout has passed
		if time.Since(cb.openTime) >= cb.config.Timeout {
			cb.transitionTo(StateHalfOpen)
			return true
		}
		return false

	case StateHalfOpen:
		// Allow limited requests through
		return cb.failures < cb.config.MaxRequests

	default:
		return false
	}
}

// recordResult records the result of a request.
func (cb *CircuitBreaker) recordResult(err error) {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if err == nil {
		cb.onSuccess()
		return
	}
	cb.onFailure()
}

func (cb *CircuitBreaker) onSuccess() {
	switch cb.state {
	case StateHalfOpen:
		cb.failures--
		if cb.failures <= 0 {
			cb.transitionToLocked(StateClosed)
		}
	case StateClosed:
		cb.failures = 0
	}
}

func (cb *CircuitBreaker) onFailure() {
	cb.lastFailure = time.Now()

	switch cb.state {
	case StateClosed:
		cb.failures++
		if cb.failures >= cb.config.MaxRequests {
			cb.transitionToLocked(StateOpen)
		}
	case StateHalfOpen:
		cb.failures++
		cb.transitionToLocked(StateOpen)
	}
}

func (cb *CircuitBreaker) transitionTo(newState State) {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.transitionToLocked(newState)
}

func (cb *CircuitBreaker) transitionToLocked(newState State) {
	cb.state = newState

	switch newState {
	case StateClosed:
		cb.failures = 0
	case StateOpen:
		cb.openTime = time.Now()
	case StateHalfOpen:
		cb.failures = 0
	}
}

// Reset manually resets the circuit breaker to closed state.
func (cb *CircuitBreaker) Reset() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.transitionToLocked(StateClosed)
}

File: template-parts/code-library/snippets/python/circuit_breaker.py
"""Circuit breaker pattern for Python.

Prevents cascading failures by stopping requests to a failing service.

Usage:
    cb = CircuitBreaker(failure_threshold=5, timeout=30)
    try:
        result = cb.call(fragile_operation)
    except CircuitBreakerOpen:
        # handle open circuit
        pass
"""

from __future__ import annotations

import asyncio
import threading
import time
from enum import Enum
from typing import Any, Callable, ParamSpec, TypeVar

P = ParamSpec("P")
T = TypeVar("T")


class CircuitState(Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half-open"


class CircuitBreakerOpen(Exception):
    """Raised when the circuit breaker is open."""
    pass


class CircuitBreaker:
    """Circuit breaker implementation.

    Args:
        failure_threshold: Number of failures before opening the circuit.
        success_threshold: Number of successes in half-open before closing.
        timeout: Seconds before attempting to close the circuit.
        expected_exceptions: Exceptions that count as failures.
    """

    def __init__(
        self,
        failure_threshold: int = 5,
        success_threshold: int = 2,
        timeout: float = 60.0,
        expected_exceptions: tuple[type[Exception], ...] = (Exception,),
    ) -> None:
        self.failure_threshold = failure_threshold
        self.success_threshold = success_threshold
        self.timeout = timeout
        self.expected_exceptions = expected_exceptions

        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._success_count = 0
        self._last_failure_time: float | None = None
        self._lock = threading.Lock()

    @property
    def state(self) -> CircuitState:
        with self._lock:
            self._check_state_transition()
            return self._state

    def _check_state_transition(self) -> None:
        """Check if state should transition based on timeout."""
        if self._state == CircuitState.OPEN and self._last_failure_time is not None:
            if time.monotonic() - self._last_failure_time >= self.timeout:
                self._state = CircuitState.HALF_OPEN
                self._success_count = 0

    def call(self, func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
        """Execute the function through the circuit breaker.

        Raises:
            CircuitBreakerOpen: If the circuit is open.
        """
        if self.state == CircuitState.OPEN:
            raise CircuitBreakerOpen("circuit breaker is open")

        try:
            result = func(*args, **kwargs)
            self._on_success()
            return result
        except self.expected_exceptions as e:
            self._on_failure()
            raise e

    def _on_success(self) -> None:
        with self._lock:
            if self._state == CircuitState.HALF_OPEN:
                self._success_count += 1
                if self._success_count >= self.success_threshold:
                    self._state = CircuitState.CLOSED
                    self._failure_count = 0
            elif self._state == CircuitState.CLOSED:
                self._failure_count = 0

    def _on_failure(self) -> None:
        with self._lock:
            self._failure_count += 1
            self._last_failure_time = time.monotonic()

            if self._state == CircuitState.HALF_OPEN:
                self._state = CircuitState.OPEN
            elif self._failure_count >= self.failure_threshold:
                self._state = CircuitState.OPEN

    def reset(self) -> None:
        """Manually reset the circuit breaker to closed state."""
        with self._lock:
            self._state = CircuitState.CLOSED
            self._failure_count = 0
            self._success_count = 0
            self._last_failure_time = None


class AsyncCircuitBreaker:
    """Async-aware circuit breaker implementation."""

    def __init__(
        self,
        failure_threshold: int = 5,
        success_threshold: int = 2,
        timeout: float = 60.0,
        expected_exceptions: tuple[type[Exception], ...] = (Exception,),
    ) -> None:
        self.failure_threshold = failure_threshold
        self.success_threshold = success_threshold
        self.timeout = timeout
        self.expected_exceptions = expected_exceptions

        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._success_count = 0
        self._last_failure_time: float | None = None
        self._lock = asyncio.Lock()

    @property
    async def state(self) -> CircuitState:
        async with self._lock:
            self._check_state_transition()
            return self._state

    def _check_state_transition(self) -> None:
        if self._state == CircuitState.OPEN and self._last_failure_time is not None:
            if time.monotonic() - self._last_failure_time >= self.timeout:
                self._state = CircuitState.HALF_OPEN
                self._success_count = 0

    async def call(self, coro: Callable[P, Any]) -> Any:
        """Execute the async function through the circuit breaker."""
        if await self.state == CircuitState.OPEN:
            raise CircuitBreakerOpen("circuit breaker is open")

        try:
            result = await coro()
            await self._on_success()
            return result
        except self.expected_exceptions as e:
            await self._on_failure()
            raise e

    async def _on_success(self) -> None:
        async with self._lock:
            if self._state == CircuitState.HALF_OPEN:
                self._success_count += 1
                if self._success_count >= self.success_threshold:
                    self._state = CircuitState.CLOSED
                    self._failure_count = 0
            elif self._state == CircuitState.CLOSED:
                self._failure_count = 0

    async def _on_failure(self) -> None:
        async with self._lock:
            self._failure_count += 1
            self._last_failure_time = time.monotonic()

            if self._state == CircuitState.HALF_OPEN:
                self._state = CircuitState.OPEN
            elif self._failure_count >= self.failure_threshold:
                self._state = CircuitState.OPEN
